xds/internal/resolver/xds_resolver_test.go

```diff
@@ -632,123 +636,297 @@ func (s) TestXDSResolverGoodServiceUpdate(t *testing.T) {
 	}
 }
 
-// TestXDSResolverRequestHash tests a case where a resolver receives a RouteConfig update
+// TestResolverRequestHash tests a case where a resolver receives a RouteConfig update
 // with a HashPolicy specifying to generate a hash. The configSelector generated should
 // successfully generate a Hash.
-func (s) TestXDSResolverRequestHash(t *testing.T) {
+func (s) TestResolverRequestHash(t *testing.T) {
 	oldRH := envconfig.XDSRingHash
 	envconfig.XDSRingHash = true
 	defer func() { envconfig.XDSRingHash = oldRH }()
 
-	xdsR, xdsC, tcc, cancel := testSetup(t, setupOpts{target: target})
-	defer xdsR.Close()
-	defer cancel()
+	mgmtServer, err := e2e.StartManagementServer(e2e.ManagementServerOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer mgmtServer.Stop()
 
-	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
-	defer cancel()
-	waitForWatchListener(ctx, t, xdsC, targetStr)
-	xdsC.InvokeWatchListenerCallback(xdsresource.ListenerUpdate{RouteConfigName: routeStr, HTTPFilters: routerFilterList}, nil)
-	waitForWatchRouteConfig(ctx, t, xdsC, routeStr)
-	// Invoke watchAPI callback with a good service update (with hash policies
-	// specified) and wait for UpdateState method to be called on ClientConn.
-	xdsC.InvokeWatchRouteConfigCallback("", xdsresource.RouteConfigUpdate{
-		VirtualHosts: []*xdsresource.VirtualHost{
-			{
-				Domains: []string{targetStr},
-				Routes: []*xdsresource.Route{{
-					Prefix: newStringP(""),
-					WeightedClusters: map[string]xdsresource.WeightedCluster{
-						"cluster_1": {Weight: 75},
-						"cluster_2": {Weight: 25},
-					},
-					HashPolicies: []*xdsresource.HashPolicy{{
-						HashPolicyType: xdsresource.HashPolicyTypeHeader,
-						HeaderName:     ":path",
+	// Create a bootstrap configuration specifying the above management server.
+	nodeID := uuid.New().String()
+	cleanup, err := xdsbootstrap.CreateFile(xdsbootstrap.Options{
+		NodeID:    nodeID,
+		ServerURI: mgmtServer.Address,
+		Version:   xdsbootstrap.TransportV3,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cleanup()
+
+	// Build an xDS resolver that uses the above bootstrap configuration
+	// Creating the xDS resolver should result in creation of the xDS client.
+	builder := resolver.Get(xdsScheme)
+	if builder == nil {
+		t.Fatalf("resolver.Get(%v) returned nil", xdsScheme)
+	}
+	const serviceName = "my-service-client-side-xds"
+	ldsName := serviceName
+	rdsName := "route-" + serviceName
+	u, err := url.Parse("xds:///" + serviceName)
+	if err != nil {
+		t.Fatal(err)
+	}
+	tcc := newTestClientConn()
+	r, err := builder.Build(resolver.Target{URL: *u}, tcc, resolver.BuildOptions{})
+	if err != nil {
+		t.Fatalf("builder.Build(%v) returned err: %v", target, err)
+	}
+	defer r.Close()
```

Comment:
should we move this block to a helper function and use it 2 more times below?

Reply:
Done. Thanks.

Reply:
Also, added a utils function to parse the URL. I suggested this in another PR, but that is stuck now on other things. So, I pulled the file into this one so that it can be used.